### DISCUSSION TITLE
Determining convexity/concavity of a function

### DISCUSSION THREAD

**Discussion by CarloNicolini** [cvxpy/cvxpy#1348] [Q&A]
Dear CVXPY folks,

I am having problems in understanding the reason why cvxpy cannot determine the convexity/concavity of the sum of elementwise logarithm of a matrix obtained as outer product of two positive variables.
Specifically my objective function contains the term here named `objective_term`.

    import cvxpy as cp
    N = 10
    theta = cp.Variable(shape=(2*N,), pos=True)
    x = theta[0:N][:,None]
    y = theta[N:][:,None]

    objective_term = cp.sum(cp.log(cp.matmul(x, cp.transpose(y))))

The matrix `cp.matmul(x, cp.transpose(y))` is indicated as `Expression(UNKNOWN, NONNEGATIVE, (10, 10))` but once I try to solve the problem:

    cp.Problem(cp.Maximize(objective_term),[]).solve(gp=True)

I get the error saying that the problem is not DGP because the expression in outer product `cp.matmul(x, cp.transpose(y))` is not DGP.

Is there a reason why the sum of log of product of all positive variables results in this error?

How can I handle this optimization problem?

**Comment by akshayka** [comment]
First, can you try upgrading CVXPY to the latest version, if you're not already using it?

The reason I ask is because I tried your code snippet, and got a different result than you.

The first thing I noticed is that `cp.matmul(x, cp.transpose(y))` is of shape `(10, 0)`, not `(10, 10)`. This is expected because `y` is of shape `(0, 1)`. You probably didn't intend that, so I think the problem may be set up incorrectly.

Second, I'm surprised you get a DGP error, because your code does not have the flag `gp=True`. Did you solve with `gp=True`?

Third, are you trying to construct a DCP program, or a DGP program? In either case you need to make sure your problem follows the DCP or DGP rulesets. These rulesets are explained in the online documentation, available here: https://www.cvxpy.org/tutorial/index.html

**Reply by CarloNicolini** [reply]
Sorry for the typo, the variable theta is 2*N elements, so both x and y are (N,1) arrays. The problem still applies though.
CVXPY version is 1.1.12 on MacOSx.

**Reply by akshayka** [reply]
No worries. I am assuming that you are passing `gp=True` to the `solve` method.

The matmul atom is log-log convex, not log-log affine, since in general the entries of the resulting expression can be posynomials.

In your case the entries are monomials, because you have an outer product.

If `N` is small in your real application, you could construct the outer product using atomic log-log affine atoms, as done in this example: https://www.cvxpy.org/examples/dgp/rank_one_nmf.html

Another thing we could do in the future is add an outer-product atom, which was recongized as log-log affine.

**Reply by CarloNicolini** [reply]
Independently of how I create the rank 1 matrix, the result is the same expression with shape (10,10)

    >>> x@y.T
    Expression(UNKNOWN, NONNEGATIVE, (10, 10))
    
    >>> cp.vstack([cp.multiply(x[:,0],y[i,0]) for i in range(N)])
    Expression(UNKNOWN, NONNEGATIVE, (10, 10))
    
    >>> cp.matmul(x,cp.transpose(y))
    Expression(UNKNOWN, NONNEGATIVE, (10, 10))

Then once I try to apply elementwise logarithm on the matrix elements, the expression become UNKNOWNS:

    >>> cp.log( cp.matmul(x,cp.transpose(y)))
    Expression(UNKNOWN, UNKNOWN, (10,10))

**Reply by akshayka** [reply]
This suggests that we have a bug in our string formatting for expressions. CVXPY correctly parses the curvature of your second line.

```
cp.vstack([cp.multiply(x[:,0],y[i,0]) for i in range(N)]).log_log_curvature
'LOG-LOG AFFINE'

cp.log(cp.vstack([cp.multiply(x[:,0],y[i,0]) for i in range(N)])).log_log_curvature
'LOG-LOG CONCAVE'
```

However, the sum of log-log concave functions does not have known curvature (sum is a log-log convex operation).

Are you solving with `gp=True`? I want to know whether you're trying to construct and solve a DGP program or a DCP program.
